[Screen Time Refactoring] Ensure donations are scoped to an identifier

jesxilin opened the pull request:
#### 47d0e9e389af1253a25b4f7339b387c260078fa5
<pre>
[Screen Time Refactoring] Ensure donations are scoped to an identifier
<a href="https://bugs.webkit.org/show_bug.cgi?id=285350">https://bugs.webkit.org/show_bug.cgi?id=285350</a>
<a href="https://rdar.apple.com/140439343">rdar://140439343</a>

Reviewed by Wenson Hsieh, Megan Gardner, and Aditya Keerthi.

Ensure donations are scope to an identifier by using identifier from WKWebsiteDataStore.
Add API tests to confirm identifiers are supported correctly with STWebpageController.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _installScreenTimeWebpageController]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(webViewForScreenTimeTests):
(TEST(ScreenTime, IdentifierNil)):
(TEST(ScreenTime, IdentifierString)):

Canonical link: <a href="https://commits.webkit.org/288472@main">https://commits.webkit.org/288472@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0374eff296b181daeb187aa152f7af02e0b96160

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83245 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2864 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37535 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88326 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34259 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85337 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2947 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10806 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64770 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22520 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86299 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2145 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75660 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45054 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/82664 "Build is in progress. Recent messages:") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2049 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29856 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33308 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73184 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30580 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89699 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10514 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7558 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73211 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10739 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71478 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72440 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16636 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15364 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1852 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12886 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10469 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15944 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10321 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13791 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12092 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -234,6 +234,11 @@
 #endif
 
 #if ENABLE(SCREEN_TIME)
+#if PLATFORM(MAC) || PLATFORM(IOS_FAMILY)
```
I think this`#if`  is redundant?

hortont424 replied:
What's this for? Who has ENABLE(SCREEN_TIME) but isn't macOS or iOS-family??

rr-codes commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -153,4 +158,57 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+        })
+    };
+
+    NSUUID * uuid = [NSUUID UUID];
```
```suggestion
    RetainPtr uuid = [NSUUID UUID];
```

pxlcoder replied:
```suggestion
    RetainPtr uuid = [NSUUID UUID];
```

whsieh replied:
```suggestion
    NSUUID *uuid = [NSUUID UUID];
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -46,6 +46,7 @@
 #import <wtf/spi/cocoa/NSObjCRuntimeSPI.h>
 
 #if ENABLE(SCREEN_TIME)
+#import <ScreenTime/STWebHistory.h>
```
Can probably be imported in the implementation file?

jesxilin replied:
I think this was left over when I was testing history stuff. So I'll just take it out for now.

rr-codes commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -153,4 +158,57 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+    };
+
+    NSUUID * uuid = [NSUUID UUID];
+    auto websiteDataStore = [WKWebsiteDataStore dataStoreForIdentifier:uuid];
```
RetainPtr instead of auto

pxlcoder replied:
```suggestion
    RetainPtr websiteDataStore = [WKWebsiteDataStore dataStoreForIdentifier:uuid];
```

rr-codes commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -153,4 +158,57 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+
+    TestWebKitAPI::Util::run(&done);
+
+    NSString * uuidString = uuid.UUIDString;
```
```suggestion
    RetainPtr uuidString = uuid.UUIDString;
```

whsieh replied:
Minor style nit -
```suggestion
    NSString *uuidString = uuid.UUIDString;
```

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -234,6 +234,11 @@
+@interface STWebpageController (Staging_138865295)
+@property (nonatomic, copy) STWebHistoryProfileIdentifier profileIdentifier;
+@end
+#endif
```
Nit - you don't need the guards here because `PLATFORM(MAC) || PLATFORM(IOS_FAMILY)` is equivalent to `PLATFORM(COCOA)`, but we're already in an `mm` file.

```suggestion
@interface STWebpageController (Staging_138865295)
@property (nonatomic, copy) STWebHistoryProfileIdentifier profileIdentifier;
@end
```

~This should also (ideally) be moved into a separate `ScreenTimeSPI.h` header.~

(I see, this is new public API, so leaving a staging declaration here seems fine. We can remove it once it's been in relevant SDKs for a while)

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -234,6 +234,11 @@
 #endif
 
 #if ENABLE(SCREEN_TIME)
+#if PLATFORM(MAC) || PLATFORM(IOS_FAMILY)
```
is this guard needed?

jesxilin replied:
Don't think so now... changing!

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -153,4 +158,57 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+    auto websiteDataStore = [WKWebsiteDataStore dataStoreForIdentifier:uuid];
+
+    RetainPtr configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
+    configuration.get().websiteDataStore = websiteDataStore;
```
```suggestion
    [configuration setWebsiteDataStore:websiteDataStore];
```

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -46,6 +46,7 @@
 #import <wtf/spi/cocoa/NSObjCRuntimeSPI.h>
 
 #if ENABLE(SCREEN_TIME)
+#import <ScreenTime/STWebHistory.h>
```
is this import needed?

jesxilin replied:
I don't think it is actually. I think I just forgot to take this out b/c I was testing stuff. Will remove!

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -153,4 +158,57 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+    for (_WKFeature *feature in [WKPreferences _features]) {
+        if ([feature.key isEqualToString:@"ScreenTimeEnabled"])
+            [preferences _setEnabled:YES forFeature:feature];
+    }
```
Maybe you can make `webViewForScreenTimeTests` take a configuration parameter, rather than repeat this code?

jesxilin replied:
That's a good idea! I was trying to figure out how to get the configuration back so I'll try this!

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -27,19 +27,23 @@
 static bool stateDidChange = false;
 
-static RetainPtr<TestWKWebView> webViewForScreenTimeTests()
+static RetainPtr<TestWKWebView> webViewForScreenTimeTests(RetainPtr<WKWebViewConfiguration> configuration)
```
```suggestion
static RetainPtr<TestWKWebView> webViewForScreenTimeTests(WKWebViewConfiguration *configuration = nil)
```

that way you don't need to update most call sites

jesxilin replied:
As in I wouldn't need the `RetainPtr configuration = adoptNS([[WKWebViewConfiguration alloc] init]);` in the tests that don't access configuration afterwards? I don't think I understand, can you help elaborate for me?

pxlcoder replied:
Correct. In those tests, you would continue calling as `webViewForScreenTimeTests()`, without passing in a parameter.

Then in `webViewForScreenTimeTests`, you only create a configuration if the parameter is `nil`.

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -27,25 +27,32 @@
 {
-    RetainPtr configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
+    if (!configuration)
+        configuration = [[WKWebViewConfiguration alloc] init];
```
This needs `adoptNS()` to prevent a memory leak

```suggestion
        configuration = adoptNS([[WKWebViewConfiguration alloc] init]).autorelease();
```

jesxilin replied:
Got it!

pxlcoder replied:
(that alone won't compile since `configuration` is a raw pointer, you also need to create a local `RetainPtr` variable)

pxlcoder replied:
oh Wenson edited to use `autorelease`, that works too!

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -27,25 +27,32 @@
     }
-    RetainPtr webView = adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 400, 300) configuration:configuration.get()]);
+    RetainPtr webView = adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 400, 300) configuration:configuration]);
     return webView;
```
(Not new to this PR, but you could just `return` the value directly)
```suggestion
    return adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 400, 300) configuration:configuration]);
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -153,4 +160,56 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+    };
+
+    RetainPtr configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
+    RetainPtr webView = webViewForScreenTimeTests(configuration.get());
```
```suggestion
    RetainPtr webView = webViewForScreenTimeTests();
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -153,4 +159,55 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
 }
 
+TEST(ScreenTime, IdentifierNil)
+{
```
```suggestion
{
    if (![PAL::getSTWebpageControllerClass() instancesRespondToSelector:@selector(setProfileIdentifier:)]
        return;
        
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -153,4 +159,55 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+}
+
+TEST(ScreenTime, IdentifierString)
+{
```
```suggestion
{
    if (![PAL::getSTWebpageControllerClass() instancesRespondToSelector:@selector(setProfileIdentifier:)]
            return;
            
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -27,26 +27,32 @@
+    return adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 400, 300) configuration:configuration]);
 }
 
 @interface STWebpageController ()
```
I think you'll have to add `@property (nonatomic, copy) STWebHistoryProfileIdentifier profileIdentifier;` in here too, so this builds on older SDKs.

jesxilin committed:
[Screen Time Refactoring] Ensure donations are scoped to an identifier
https://bugs.webkit.org/show_bug.cgi?id=285350
rdar://140439343

Reviewed by Wenson Hsieh, Megan Gardner, and Aditya Keerthi.

Ensure donations are scope to an identifier by using identifier from WKWebsiteDataStore.
Add API tests to confirm identifiers are supported correctly with STWebpageController.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _installScreenTimeWebpageController]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(webViewForScreenTimeTests):
(TEST(ScreenTime, IdentifierNil)):
(TEST(ScreenTime, IdentifierString)):

Canonical link: https://commits.webkit.org/288472@main
